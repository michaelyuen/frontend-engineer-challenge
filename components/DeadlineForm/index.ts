export * from "./DeadlineForm";
export * from "./styles";
