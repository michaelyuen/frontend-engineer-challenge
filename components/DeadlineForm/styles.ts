import styled from "styled-components";
import { Form } from "../Form";
import { breakpoints } from "../../styles/breakpoints";

export const DeadlineFormContainer = styled(Form)`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  max-width: ${breakpoints.sm};

  h1 {
    margin-bottom: 1em;
    text-align: center;
  }

  label {
    margin-bottom: 0.5em;
  }
`;
