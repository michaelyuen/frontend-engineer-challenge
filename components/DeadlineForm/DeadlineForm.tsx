import { useEffect, useState } from "react";
import { DeadlineFormContainer } from "./styles";
import { Button, ButtonProps } from "../Button";
import { FormProps } from "../Form";
import { Input, InputProps } from "../Input";
import { isPast } from "../../utils";

/**
 * Using TypeScript Template Literal Type: https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html
 * Expected format: 2024-10-16T17:47
 */
export type DeadlineValue =
  `${number}${number}${number}${number}-${number}${number}-${number}${number}T${number}${number}:${number}${number}`;
export type DeadlineFormData = {
  deadlineName: string;
  deadlineValue: DeadlineValue;
};

export interface DeadlineFormProps extends DeadlineFormData {
  buttonProps?: ButtonProps;
  error?: string;
  onChange: InputProps["onChange"];
  onSubmit: FormProps["onSubmit"];
}

export const DeadlineForm: React.FC<DeadlineFormProps> = ({
  buttonProps,
  deadlineName,
  deadlineValue,
  error,
  onChange,
  onSubmit,
  ...props
}) => {
  const [deadlineValueError, setDeadlineValueError] = useState<string>();

  useEffect(() => {
    const isInPast = isPast(deadlineValue);
    if (isInPast) {
      setDeadlineValueError("Please select a date and time in the future");
    } else {
      setDeadlineValueError("");
    }
  }, [deadlineValue]);

  const isDisabled = !(deadlineName && deadlineValue) || !!deadlineValueError;

  return (
    <DeadlineFormContainer
      aria-label="Deadline Form"
      name="Deadline Form"
      noValidate
      onSubmit={onSubmit}
      {...props}
    >
      <h1>Deadline Countdown</h1>

      <label htmlFor="deadline-name">What&apos;s the deadline?</label>
      <Input
        aria-label="Enter name for deadline"
        aria-required="true"
        id="deadline-name"
        name="deadlineName"
        onChange={onChange}
        required
        value={deadlineName}
      />

      <label htmlFor="deadline-date-time">When is it?</label>
      <Input
        aria-label="Enter date and time for deadline"
        aria-required="true"
        error={deadlineValueError}
        id="deadline-date-time"
        min={new Date()
          .toLocaleString("sv", { dateStyle: "short", timeStyle: "short" })
          .replace(" ", "T")}
        name="deadlineValue"
        onChange={onChange}
        required
        type="datetime-local"
        value={deadlineValue}
      />
      <Button disabled={isDisabled} type="submit" {...buttonProps}>
        Start the countdown!
      </Button>
    </DeadlineFormContainer>
  );
};
