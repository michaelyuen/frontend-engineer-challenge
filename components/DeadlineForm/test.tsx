import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import { DeadlineForm } from "./DeadlineForm";

describe("DeadlineForm", () => {
  test("renders", () => {
    render(
      <DeadlineForm
        deadlineName=""
        deadlineValue=""
        onChange={jest.fn}
        onSubmit={jest.fn}
      />
    );
    expect(screen.getByText("Deadline Countdown")).toBeInTheDocument();
    expect(screen.getByLabelText("Deadline Form")).toBeInTheDocument();
    expect(
      screen.getByLabelText("Enter name for deadline")
    ).toBeInTheDocument();
    expect(
      screen.getByLabelText("Enter date and time for deadline")
    ).toBeInTheDocument();
    expect(screen.getByText("Start the countdown!")).toBeInTheDocument();
  });

  /**
   * Changes every time due to deadlineValue's min
   */
  xtest("matches snapshot", () => {
    const { container } = render(
      <DeadlineForm
        deadlineName=""
        deadlineValue=""
        onChange={jest.fn}
        onSubmit={jest.fn}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("has no a11y violations", async () => {
    const { container } = render(
      <DeadlineForm
        deadlineName=""
        deadlineValue=""
        onChange={jest.fn}
        onSubmit={jest.fn}
      />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});
