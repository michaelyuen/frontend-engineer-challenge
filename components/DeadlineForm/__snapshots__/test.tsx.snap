// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeadlineForm matches snapshot 1`] = `
<div>
  <form
    aria-label="Deadline Form"
    class="sc-bcXHqe sc-gswNZR iyTodM"
    name="Deadline Form"
    novalidate=""
  >
    <h1>
      Deadline Countdown
    </h1>
    <label
      for="deadline-name"
    >
      What's the deadline?
    </label>
    <input
      aria-label="Enter name for deadline"
      aria-required="true"
      class="sc-pyfCe dfbbuZ"
      id="deadline-name"
      name="deadlineName"
      required=""
      type="text"
      value=""
    />
    <label
      for="deadline-date-time"
    >
      When is it?
    </label>
    <input
      aria-label="Enter date and time for deadline"
      aria-required="true"
      class="sc-pyfCe dfbbuZ"
      id="deadline-date-time"
      min="2022-10-16T22:28"
      name="deadlineValue"
      required=""
      type="datetime-local"
      value=""
    />
    <button
      class="sc-iBYQkv MerZa"
      disabled=""
      type="submit"
    >
      Start the countdown!
    </button>
  </form>
</div>
`;
